### PR TITLE
Ubuntu based docker image setup and docker CI setup

### PR DESCRIPTION
**Pull request by yashaswi2000**
## Description
Dockerfiles for building the lind containers with Ubuntu based image.
CI update to trigger building and pushing latest versions of Lind to Dockerhub.

#### Key Files, ignore rest:
- .github/workflows/daily-builds.yml
- src/ubuntu-docker/lind-base/Dockerfile
- src/ubuntu-docker/lind-glibc/Dockerfile
- src/ubuntu-docker/lind-full/Dockerfile
- src/mklind

### Steps to create CI container:
- build lind-full using the following command:
```
docker build --build-arg ENV=ci --build-arg docker_gid=999 -t securesystemslab/lind_ubuntu:lind-full src/ubuntu-docker/lind-full/
```
here when ENV is pass it creates a docker out of docker capable image, where docker_gid is the group id of docker group in host system

- to run this new image:
```
docker run -v /var/run/docker.sock:/var/run/docker.sock --privileged --ipc=host --init --cap-add=SYS_PTRACE -it securesystemslab/lind_ubuntu:lind-full /bin/bash
```
This mounts the host docker into the container, letting us to call docker enginer from container.

### steps to create dev/deployment container:

```
docker run  --privileged --ipc=host --init --cap-add=SYS_PTRACE -it securesystemslab/lind_ubuntu:lind-full /bin/bash
```

Fixes # (issue)

 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- Tested Docker ouf of Docker feature to build and push images.
- make test runs without failures
- make core-utils runs without failures

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, safeposix-rust)

### CODE REVIEW

File: src/docker/lind-full/Dockerfile [hunk 1]
```diff
@@ -27,14 +27,14 @@ USER lind
 
 WORKDIR /home/lind/lind_project
 
-# finish install
-RUN git pull -t -j8
-RUN ./src/mklind rustposix
-RUN ./src/mklind nacl
-RUN ./src/mklind install
+# # finish install
```
Comment (rennergade) [inline src/docker/lind-full/Dockerfile]: why is this commented out?

**Review by rennergade — CHANGES_REQUESTED** [review]
I think this PR needs a much lengthier explanation since I cant really discern what to actually look at. We probably shouldn't commit things that are commented out though.

Would love @namanlalitnyu to look through this before I do.

File: src/docker/lind-base/Dockerfile [hunk 2]
```diff
@@ -92,6 +92,7 @@ WORKDIR /home/lind/
 RUN rm -rf "/home/lind/build/" "/root/build/" "/var/cache/pacman/pkg/"
 RUN chown -Rc lind:lind .
 
+
```
Comment (namanlalitnyu) [inline src/docker/lind-base/Dockerfile]: I think we can remove this extra line.

File: src/docker/lind-full/Dockerfile (hunk 1 above)
Comment (namanlalitnyu) [reply src/docker/lind-full/Dockerfile]: Same doubt.

File: src/ubuntu-docker/lind-base/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    build-essential \
+    git mercurial rsync perl lynx ltrace \
+	w3m gawk procps ruby strace gdb reptyr ed \
+  	vim elfutils libelf-dev gdb less \
+  	netcat htop bc ctags \
+ 	libncurses-dev gcc-multilib gperf diffutils \
+	google-perftools texinfo fortune-mod gnupg highlight fasd \
+	python2.7 python2.7-dev p7zip fuse ack \
+	isc-dhcp-client iputils-ping iproute2 \
+ 	net-tools openresolv openssh-server subversion \
```
Comment (namanlalitnyu) [inline src/ubuntu-docker/lind-base/Dockerfile:36]: Just curious, how do we know these are the libraries that we need for this Docker container?

File: src/ubuntu-docker/lind-base/etc/tmux.conf [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+set -g status-interval 1
+set -g status-attr bright
+set -g status-fg white
+set -g status-bg black
+set -g status-left-length 20
+set -g status-left "∙ #[fg=colour240,bg=white] %b %d %Y #[bg=colour240,fg=white] #(tmux-mem-cpu-load) #[default]"
+set -g status-right '#(~/bin/tmux_battery_charge_indicator.sh) #[bg=colour240,fg=white] %I:%M:%S %P #[bg=white,fg=colour240] %a '
+set -g status-justify centre
+set -g window-status-current-format '#[fg=green,bg=black](#I.#P|#F|#W)#[default]'
+set -g window-status-format '#[fg=lightcyan,bg=black]#I.#P#F#W'
```
Comment (namanlalitnyu) [inline src/ubuntu-docker/lind-base/etc/tmux.conf:231]: Is this an auto generated file?

**Review by namanlalitnyu — CHANGES_REQUESTED** [review]
We have a lot of extra files being generated here, which I think are the dependencies for "make-3.82" library. Do we need them or should we add them in the .gitignore?

File: src/docker/lind-base/Dockerfile (hunk 2 above)
Comment (yashaswi2000) [reply src/docker/lind-base/Dockerfile]: removed

File: src/ubuntu-docker/lind-base/Dockerfile (hunk 3 above)
Comment (yashaswi2000) [reply src/ubuntu-docker/lind-base/Dockerfile:36]: we have referenced them from the arch linux image and what libraries it was using. although we could do a analysis and figure out some redundancies to reduce image size.

File: src/ubuntu-docker/lind-base/etc/tmux.conf (hunk 4 above)
Comment (yashaswi2000) [reply src/ubuntu-docker/lind-base/etc/tmux.conf:231]: no this again is an existing tmux config we were using in arch linux

File: src/ubuntu-docker/lind-base/Dockerfile (hunk 3 above)
Comment (namanlalitnyu) [reply src/ubuntu-docker/lind-base/Dockerfile:36]: Got it.

**Comment by rennergade** [comment]
I need you to point out specific parts for me to review since I can't really look at all these files.

**Comment by rennergade** [comment]
A list of testing you've done so far would be good too.

**Comment by yashaswi2000** [comment]
added the files to review in main description and the types of tests that are passed.

**Review by rennergade — APPROVED** [review]
Ok this looks good!

**Pull request merged**